### PR TITLE
[New feature] add transformer layer

### PR DESCRIPTION
**Pull request by zhangxiaoli73**
## What changes were proposed in this pull request?

Add transformer layer to support language model case and translation case

## How was this patch tested?

unit tests (comparing with results from tf)

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/XXX

### CODE REVIEW

File: pyspark/bigdl/nn/layer.py [hunk 1]
```diff
@@ -964,6 +964,20 @@ class ExpandSize(Layer):
     def __init__(self, sizes, bigdl_type="float"):
         super(ExpandSize, self).__init__(None, bigdl_type, sizes)
 
+class TransformerLayer(Layer):
+
+    '''
+    Implementation TransformerLayer
```
Comment (wzhongyuan) [inline pyspark/bigdl/nn/layer.py]: why we should add `Layer` to `Transformer`, it's actually under nn right ?

Comment (wzhongyuan) [reply pyspark/bigdl/nn/layer.py]: and `Implementation for` or just put `Implementation` after the layer name

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Attention.scala [hunk 2]
```diff
@@ -43,11 +43,11 @@ class Attention[T: ClassTag](
 
     // Layers for linearly projecting the queries, keys, and values.
     val queryLayer = TransformerOperation.dense(
-      hiddenSize, hiddenSize, false, name = "q").inputs(inputX)
+      hiddenSize, hiddenSize, false, name = s"${this.getName()}_q").inputs(inputX)
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Attention.scala]: why the change ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/CAddTable.scala [hunk 3]
```diff
@@ -47,12 +47,10 @@ class CAddTable[T: ClassTag, D: ClassTag](val inplace: Boolean = false)(
   private var bufferSumOutput: Tensor[D] = null
 
   private def expandWithDims(smallSize: Array[Int], otherSize: Array[Int]): Boolean = {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/CAddTable.scala]: `smallSize` and `otherSize` looks confusing.

Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/CAddTable.scala]: also the method returns boolean.  probably name as `canExpand(inputSize: Array[Int],  targetSize: Array[Int])`

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/FeedForwardNetwork.scala [hunk 4]
```diff
@@ -36,12 +36,13 @@ class FeedForwardNetwork[T: ClassTag](val hiddenSize: Int, val filterSize: Int,
   override def buildModel(): Module[T] = {
     val input = Input()
     val filterLayer = TransformerOperation.dense(
-      hiddenSize, filterSize, bias = true, activation = ReLU[T]()).inputs(input)
+      hiddenSize, filterSize, bias = true, activation = ReLU[T](),
+      name = s"${this.getName()}_filter_layer").inputs(input)
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/FeedForwardNetwork.scala]: why add module name ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+class TransformerLayer[T: ClassTag](
+   val vocabSize: Int,
+   val hiddenSize: Int,
+   val numHeads: Int,
+   val filterSize: Int,
+   val numHiddenlayers: Int,
+   val postprocessDropout: Float,
+   val attentionDropout: Float,
+   val reluDropout: Float,
+   val problem: ProblemType = LanguageModel)
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: problem looks confusing,  just name it  type ?

Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: `val type: TransformerType = LanguageModel`

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+    val constantValue = math.sqrt(hiddenSize)
+    val embeddingInput = MulConstant(constantValue).inputs(embedding.inputs(inNode))
+    val embeddingOutput = MulConstant(constantValue).inputs(embeddingDupicate.inputs(tarNode))
+    val encoder_outputs = encode(embeddingInput, attention_bias)
+    val outNode = decode(embeddingOutput, encoder_outputs, attention_bias)
+    Graph(Array(inNode, tarNode), outNode)
+  }
+
+  private def buildLM(): Module[T] = {
+    val inNode = Input()
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: inputNode ? we I assume we name similar in other code as well

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+    val constantValue = math.sqrt(hiddenSize)
+    val embeddingInput = MulConstant(constantValue).inputs(
+      LookupTable[T](vocabSize, hiddenSize).inputs(inNode))
+    val outNode = decode(embeddingInput)
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: outputNode ? we I assume we name similar in other code as well

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+    problem match {
+      case LanguageModel => buildLM()
+      case Translation => buildTranslation()
+    }
+  }
+
+  // inputs: int tensor with shape [batch_size, input_length].
+  // targets: None or int tensor with shape [batch_size, target_length].
+  private def buildTranslation(): Module[T] = {
+    val inNode = Input()
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: inputNode ? we I assume we name similar in other code as well

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+    val params1 = embedding.getParameters()
+    val params2 = embeddingDupicate.getParameters()
+    params1._1.set(params2._1)
+    params1._2.set(params2._2)
+
+    val constantValue = math.sqrt(hiddenSize)
+    val embeddingInput = MulConstant(constantValue).inputs(embedding.inputs(inNode))
+    val embeddingOutput = MulConstant(constantValue).inputs(embeddingDupicate.inputs(tarNode))
+    val encoder_outputs = encode(embeddingInput, attention_bias)
+    val outNode = decode(embeddingOutput, encoder_outputs, attention_bias)
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: outputNode ? we I assume we name similar in other code as well

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala [hunk 10]
```diff
@@ -0,0 +1,333 @@
...
+  private def buildLM(): Module[T] = {
+    val inNode = Input()
+    val constantValue = math.sqrt(hiddenSize)
+    val embeddingInput = MulConstant(constantValue).inputs(
+      LookupTable[T](vocabSize, hiddenSize).inputs(inNode))
+    val outNode = decode(embeddingInput)
+    Graph(inNode, outNode)
+  }
+
+  private[nn] def encode(inputs: ModuleNode[T], attention_bias: ModuleNode[T]): ModuleNode[T] = {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: attentionBias

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala [hunk 11]
```diff
@@ -0,0 +1,333 @@
...
+  (implicit ev: TensorNumeric[T]) extends BaseModule[T] {
+
+  override def buildModel(): Module[T] = {
+    problem match {
+      case LanguageModel => buildLM()
+      case Translation => buildTranslation()
+    }
+  }
+
+  // inputs: int tensor with shape [batch_size, input_length].
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: where is inputs

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala [hunk 12]
```diff
@@ -0,0 +1,333 @@
...
+  // targets: None or int tensor with shape [batch_size, target_length].
+  private def buildTranslation(): Module[T] = {
+    val inNode = Input()
+    val tarNode = Input()
+    val attention_bias = new AttentionBiasConstant().inputs(inNode)
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: attentionBias

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala [hunk 13]
```diff
@@ -0,0 +1,333 @@
...
+
+    val embedding = LookupTable[T](vocabSize, hiddenSize)
+    val embeddingDupicate = LookupTable[T](vocabSize, hiddenSize)
+    // parameter share
+    val params1 = embedding.getParameters()
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: params1 is not a meaningful name, also for params2

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala [hunk 14]
```diff
@@ -0,0 +1,333 @@
...
+    val params2 = embeddingDupicate.getParameters()
+    params1._1.set(params2._1)
+    params1._2.set(params2._2)
+
+    val constantValue = math.sqrt(hiddenSize)
+    val embeddingInput = MulConstant(constantValue).inputs(embedding.inputs(inNode))
+    val embeddingOutput = MulConstant(constantValue).inputs(embeddingDupicate.inputs(tarNode))
+    val encoder_outputs = encode(embeddingInput, attention_bias)
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: encoderOutput ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala [hunk 15]
```diff
@@ -0,0 +1,333 @@
...
+    val embeddingInput = MulConstant(constantValue).inputs(
+      LookupTable[T](vocabSize, hiddenSize).inputs(inNode))
+    val outNode = decode(embeddingInput)
+    Graph(inNode, outNode)
+  }
+
+  private[nn] def encode(inputs: ModuleNode[T], attention_bias: ModuleNode[T]): ModuleNode[T] = {
+    // Prepare inputs to the layer stack by adding positional encodings and
+    // applying dropout.
+    val input2 = new EncodePositionConstant().inputs(inputs)
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: input2 is not good name

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala [hunk 16]
```diff
@@ -0,0 +1,333 @@
...
+    val encoder_inputs = CAddTable().inputs(inputs, input2)
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: naming

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala [hunk 17]
```diff
@@ -0,0 +1,333 @@
...
+    val decoder_input_drop = if (train) {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: naming

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala [hunk 18]
```diff
@@ -0,0 +1,333 @@
...
+      val postDropOut = Dropout(1- postprocessDropout)
+      postDropOut.inputs(encoder_inputs)
+    } else encoder_inputs
+
+    encodeStack(numHiddenlayers, decoder_input_drop, attention_bias)
+  }
+
+  private[nn] def decode(targets: ModuleNode[T],
+                     encoder_outputs: ModuleNode[T] = null,
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: naming

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala [hunk 19]
```diff
@@ -0,0 +1,333 @@
...
+                     attention_bias: ModuleNode[T] = null): ModuleNode[T] = {
+    val decoder_input = new TransformerPrepareDecoder().inputs(targets)
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: naming, please check below similar ones

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala [hunk 20]
```diff
@@ -0,0 +1,333 @@
...
+                              attention_bias: ModuleNode[T]): ModuleNode[T] = {
+    decodeStack(num_layers, encoder_input, attention_bias, preName = "encode")
+  }
+
+  private[nn] def decodeStack(num_layers: Int,
+                              decoder_input: ModuleNode[T],
+                              decoder_self_attention_bias: ModuleNode[T],
+                              encoder_outputs: ModuleNode[T] = null,
+                              attention_bias: ModuleNode[T] = null,
+                              preName: String = "decode"): ModuleNode[T] = {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: what does pre here mean ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala [hunk 21]
```diff
@@ -0,0 +1,333 @@
...
+    output
+  }
+
+  override def updateGradInput(input: Tensor[T], gradOutput: Tensor[T]): Tensor[T] = {
+    gradInput.resizeAs(input).zero()
+    gradInput
+  }
+}
+
+private[nn] class TransformerPrepareDecoder[T: ClassTag](implicit ev: TensorNumeric[T])
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: is this part of preprcessing ?

File: pyspark/bigdl/nn/layer.py (hunk 1 above)
Comment (zhangxiaoli73) [reply pyspark/bigdl/nn/layer.py]: ok, change to "Implementation for".
As our preprocessing is called "Transformer", so here "TransformerLayer" to separate from preprocessing.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Attention.scala (hunk 2 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Attention.scala]: a model may have multi attention layers. If names of two layers are same, `getParametersTable` will only get one.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/CAddTable.scala (hunk 3 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/CAddTable.scala]: ok, changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/FeedForwardNetwork.scala (hunk 4 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/FeedForwardNetwork.scala]: We have to copy tf weights to ours when doing unit tests.
If no name, it is hard to match correctly.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala (hunk 5 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: good, changed.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala (hunk 6 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: ok, changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala (hunk 7 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: ok, changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala (hunk 8 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala (hunk 9 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala (hunk 21 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: yes. It is encode position.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala (hunk 15 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: ok, changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala (hunk 19 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: ok, all changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala (hunk 20 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: `preName` will be used in encode and decode to distinguish layers.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala (hunk 21 above)
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: why not just name it `Decoder` as it's inside the same class ?

Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: Now, I named it `PositionEncodeWithShift` as it is shift input and then do position encode.

File: pyspark/bigdl/nn/layer.py (hunk 1 above)
Comment (wzhongyuan) [reply pyspark/bigdl/nn/layer.py]: it should be fine as it's under different package. From `nn` package's perspective, no need to add `Layer` explicitly

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala [hunk 22]
```diff
@@ -0,0 +1,366 @@
...
+import com.intel.analytics.bigdl.tensor.{Storage, Tensor}
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.utils.{T, Table}
+
+import scala.collection.mutable.ArrayBuffer
+import scala.reflect.ClassTag
+
+/**
+ * Transformer model from "Attention Is All You Need".
+ * The Transformer model consists of an encoder and a decoder. Both are stacks
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: a decoder. Both are ->  a decoder, both are

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala [hunk 23]
```diff
@@ -0,0 +1,366 @@
...
+   val filterSize: Int,
+   val numHiddenlayers: Int,
+   val postprocessDropout: Float,
+   val attentionDropout: Float,
+   val reluDropout: Float,
+   val problem: TransformerType = LanguageModel)
+  (implicit ev: TensorNumeric[T]) extends BaseModule[T] {
+
+  override def buildModel(): Module[T] = {
+    problem match {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: type ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala [hunk 24]
```diff
@@ -0,0 +1,366 @@
...
+                              encoderInput: ModuleNode[T],
+                              attentionBias: ModuleNode[T]): ModuleNode[T] = {
+    decodeStack(numLayers, encoderInput, attentionBias, preName = "encode")
+  }
+
+  private[nn] def decodeStack(numLayers: Int,
+                              decoderInput: ModuleNode[T],
+                              decoderSelfAttentionBias: ModuleNode[T],
+                              encoderOutputs: ModuleNode[T] = null,
+                              attentionBias: ModuleNode[T] = null,
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: what's the difference with `decoderSelfAttentionBias`  in terms of naming ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala [hunk 25]
```diff
@@ -0,0 +1,366 @@
...
+  }
+
+
+  private[nn] def encodeStack(numLayers: Int,
+                              encoderInput: ModuleNode[T],
+                              attentionBias: ModuleNode[T]): ModuleNode[T] = {
+    decodeStack(numLayers, encoderInput, attentionBias, preName = "encode")
+  }
+
+  private[nn] def decodeStack(numLayers: Int,
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: seems this method is shared by both encoding and decoding, please rename to a common one, and make the `preName` as 'xxxType'

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala [hunk 26]
```diff
@@ -0,0 +1,366 @@
...
+                              decoderInput: ModuleNode[T],
+                              decoderSelfAttentionBias: ModuleNode[T],
+                              encoderOutputs: ModuleNode[T] = null,
+                              attentionBias: ModuleNode[T] = null,
+                              preName: String = "decode"): ModuleNode[T] = {
+    var input = decoderInput
+    var i = 0
+    while (i < numLayers) {
+      val selfAttention = new Attention[T](hiddenSize, numHeads, attentionDropout)
+      val selfAttentionModel = prePostProcessingSelfAttention(
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: prePost ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala [hunk 27]
```diff
@@ -0,0 +1,366 @@
...
+        selfAttention, input, decoderSelfAttentionBias,
+        s"${preName}_self_attention_${i}")
+      input = selfAttentionModel
+
+      if (encoderOutputs != null && attentionBias != null) {
+        val encdecAttention = new Attention[T](hiddenSize, numHeads, attentionDropout)
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: encdec? typo ？

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala [hunk 28]
```diff
@@ -0,0 +1,366 @@
...
+      }
+
+      val ffn = new FeedForwardNetwork[T](hiddenSize, filterSize, reluDropout)
+      val ffnModel = prePostProcessingFFN(ffn, input, s"${preName}_ffn_${i}")
+      input = ffnModel
+
+      i += 1
+    }
+    val norm = new LayerNormalization[T](hiddenSize).inputs(input)
+    norm
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: just consolidate the 2 lines into one ?
```scala
new LayerNormalization[T](hiddenSize).inputs(input)
```

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala [hunk 29]
```diff
@@ -0,0 +1,366 @@
...
+  }
+
+  private def prePostProcessingSelfAttention(layer: Module[T], decoderInput: ModuleNode[T],
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: `prePost` is confusing

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala [hunk 30]
```diff
@@ -0,0 +1,366 @@
...
+    decoderSelfAttentionBias: ModuleNode[T], preName: String): ModuleNode[T] = {
+    val norm = new LayerNormalization[T](hiddenSize).setName(preName + "/norm")
+        .inputs(decoderInput)
+    val drop = Dropout[T](1 - postprocessDropout).setName(preName + "/dropout")
+        .inputs(layer.setName(preName + "/self_attention")
+        .inputs(norm, norm, decoderSelfAttentionBias))
+    CAddTable().inputs(decoderInput, drop)
+  }
+
+  private def prePostProcessingEncDecAttention(
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: `prePost` is confusing

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala [hunk 31]
```diff
@@ -0,0 +1,366 @@
...
+    val norm = new LayerNormalization[T](hiddenSize).setName(preName + "/norm")
+      .inputs(decoderInput)
+    val drop = Dropout[T](1 - postprocessDropout).setName(preName + "/dropout")
+      .inputs(layer.setName(preName + "/encdec_attention")
+        .inputs(norm, encoderOutput, attentionBias))
+    CAddTable().inputs(decoderInput, drop)
+  }
+
+  private def prePostProcessingFFN(layer: Module[T],
+    decoderInput: ModuleNode[T], preName: String): ModuleNode[T] = {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: maybe change `preName` to `processType` ？

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala [hunk 32]
```diff
@@ -0,0 +1,366 @@
...
+  def apply[T: ClassTag](
+     vocabSize: Int,
+     hiddenSize: Int,
+     numHeads: Int,
+     filterSize: Int,
+     numHiddenlayers: Int,
+     postprocessDropout: Float,
+     attentionDropout: Float,
+     reluDropout: Float,
+     problem: TransformerType = LanguageModel)
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: type ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala [hunk 33]
```diff
@@ -0,0 +1,366 @@
...
+   * @return
+   */
+  private def attentionBiasLowerTriangle[T: ClassTag](
+    length: Int, output: Tensor[T])(implicit ev: TensorNumeric[T]): Tensor[T] = {
+    val arr = output.storage().array()
+    for (i <- 0 to (length - 1)) {
+      var j = length - 1
+      while (j > i) {
+        // reminder: here not 1
+        arr(i * length + j) = ev.fromType(-1e9)
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: Any more efficient way for setting value with constant

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala (hunk 22 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: ok, changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala (hunk 24 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: For translation， we have to feed input and target to model when training.
`decoderSelfAttentionBias` generated from target,  and `attentionBias` from input.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala (hunk 29 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: Both pre processing and post processing.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala (hunk 33 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: ok,changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala (hunk 28 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: ok, changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala (hunk 27 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: no, it means attention with encode output and decode input

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala (hunk 31 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: preName is added as part of layer name, e.g. `preName` + `realName` 
so I think `preName` is more proper.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala (hunk 23 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: ok, change to "transformerType" as "type" is keyword.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala (hunk 25 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: ok, change to block, and make `preName` to 'blockType', which can be "encode" or "decode"

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala (hunk 24 above)
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: but we cannot see from their names one is from input and one is from target, need to make it more explicit

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala (hunk 29 above)
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: then just name `processSelfAttention` ?

File: pyspark/bigdl/nn/layer.py (hunk 1 above)
Comment (zhangxiaoli73) [reply pyspark/bigdl/nn/layer.py]: ok, changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala (hunk 29 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala]: good, changed

**Comment by zhangxiaoli73** [comment]
jenkins pass: 
http://mlt-bddl01.sh.intel.com:8080/job/BigDL-PRVN-lite/620/
